fdbclient/NativeAPI.actor.cpp

```diff
@@ -3348,74 +3348,28 @@ void enableClientInfoLogging() {
 	TraceEvent(SevInfo, "ClientInfoLoggingEnabled");
 }
 
-ACTOR Future<Void> snapshotDatabase(Reference<DatabaseContext> cx, StringRef snapPayload, UID snapUID, Optional<UID> debugID) {
-	TraceEvent("SnapshotDatabaseEnter")
-		.detail("SnapPayload", snapPayload)
-		.detail("SnapUID", snapUID);
-	try {
-		if (debugID.present()) {
-			g_traceBatch.addEvent("TransactionDebug", debugID.get().first(), "NativeAPI.snapshotDatabase.Before");
-		}
-
-		choose {
-			when(wait(cx->onMasterProxiesChanged())) { throw operation_failed(); }
-			when(wait(loadBalance(cx->getMasterProxies(false), &MasterProxyInterface::proxySnapReq, ProxySnapRequest(snapPayload, snapUID, debugID), cx->taskID, true /*atmostOnce*/ ))) {
-				if (debugID.present())
-					g_traceBatch.addEvent("TransactionDebug", debugID.get().first(),
-											"NativeAPI.SnapshotDatabase.After");
-			}
-		}
-	} catch (Error& e) {
-		TraceEvent("SnapshotDatabaseError")
-			.error(e)
-			.detail("SnapPayload", snapPayload)
-			.detail("SnapUID", snapUID);
-		throw;
-	}
-	return Void();
-}
-
-ACTOR Future<Void> snapCreate(Database cx, StringRef snapCmd, UID snapUID) {
-	// remember the client ID before the snap operation
-	state UID preSnapClientUID = cx->clientInfo->get().id;
-
+ACTOR Future<Void> snapCreate(Database cx, Standalone<StringRef> snapCmd, UID snapUID) {
 	TraceEvent("SnapCreateEnter")
 	    .detail("SnapCmd", snapCmd.toString())
-	    .detail("UID", snapUID)
-	    .detail("PreSnapClientUID", preSnapClientUID);
-
-	StringRef snapCmdArgs = snapCmd;
-	StringRef snapCmdPart = snapCmdArgs.eat(":");
-	Standalone<StringRef> snapUIDRef(snapUID.toString());
-	Standalone<StringRef> snapPayloadRef = snapCmdPart
-		.withSuffix(LiteralStringRef(":uid="))
-		.withSuffix(snapUIDRef)
-		.withSuffix(LiteralStringRef(","))
-		.withSuffix(snapCmdArgs);
+	    .detail("UID", snapUID);
 
 	try {
-		Future<Void> exec = snapshotDatabase(Reference<DatabaseContext>::addRef(cx.getPtr()), snapPayloadRef, snapUID, snapUID);
-		wait(exec);
+		loop {
+			choose {
+				when(wait(cx->onMasterProxiesChanged())) {}
+				when(wait(loadBalance(cx->getMasterProxies(false), &MasterProxyInterface::proxySnapReq, ProxySnapRequest(snapCmd, snapUID, snapUID), cx->taskID, true /*atmostOnce*/ ))) {
```

Comment:
It seems strange to pass `true /*atmostOnce*/` to loadBalance but also be in a retry loop. Maybe we could wait for `DatabaseContext::getMasterProxiesFuture` before attempting the snapshot, and then not retry it?

Reply:
I like this idea - it makes the intent much clearer

Reply:
I prefer this solution too. per-se, retrying with snapshot does not lead to incorrect results, but it has side affects like leaving some extra disk snapshots that needs to be cleaned up etc.

Reply:
With this change, number of snap tests fail. In the cases where the client retry with new snapshot creates, the call get struck in loadBalance call and the tests timeout, need to debug why it is struck in there. Until then, reverting this commit 943b2f2.

Reply:
Maybe we still want the
```
choose {
    when(wait(cx->onMasterProxiesChanged())) {}
    ...
}
```

That way if the master proxies change the snapshot attempt is cancelled instead of hanging forever

Reply:
My understanding is that it should not hang in the first place, because either I get a response or the failure monitor should kick in. Let me look at it more.